### PR TITLE
#11 PR - Add initial requirement tests

### PR DESCRIPTION
**Pull request by jonmanahan**
This PR covers [ticket #11](https://github.com/jramnani/GildedRoseJavaLLC/issues/11)

Adds gitignore file

Adds tests to cover shared simple cases:

- All items have a SellIn value which denotes the number of days we have to sell the item
- All items have a Quality value which denotes how valuable the item is
- At the end of each day our system lowers both values for every item

Adds tests to cover shared interesting edge case bullets:

- Once the sell by date has passed, Quality degrades twice as fast
- The Quality of an item is never negative
- "Aged Brie" actually increases in Quality the older it gets
- The Quality of an item is never more than 50
- "Sulfuras", being a legendary item, never has to be sold or decreases in Quality
- "Backstage passes", like aged brie, increases in Quality as its SellIn value approaches;
Quality increases by 2 when there are 10 days or less and by 3 when there are 5 days or less but
Quality drops to 0 after the concert

### CODE REVIEW

**Comment by elijahsullivan** [comment]
How do I run this locally. What prerequisite software do I need installed on my machine, if any?

**Comment by riccjohn** [comment]
Suggestion: On the right side of the PR, if you add this to the correct Project, it'll show up in the board. I'd also give this a label of "Task" just so you get used to how (some) client teams will handle different ticket types.

**Comment by riccjohn** [comment]
~Question: Are there some files that should be git-ignored here?~ Never mind! I see the diff is actually coming from you removing them from git.

File: src/test/java/com/gildedrose/GildedRoseTest.java [hunk 1]
```diff
@@ -3,15 +3,105 @@
...
+    void qualityIsNeverMoreThan50() {
+        Item[] items = new Item[] { new Item("Aged Brie", 0, 50) };
+        GildedRose app = new GildedRose(items);
+        app.updateQuality();
+        assertTrue(app.items[0].quality <= 50);
+    }
+
+    @Test
+    void qualityDegradesTwiceAsFastWhenPassedSellBy() {
+        Item[] items = new Item[] { new Item("FLAvorTown", 0, 50) };
```
Comment (mrfishball) [inline src/test/java/com/gildedrose/GildedRoseTest.java]: What is the value for `SellBy`?

File: src/test/java/com/gildedrose/GildedRoseTest.java [hunk 2]
```diff
@@ -3,15 +3,105 @@
...
         Item[] items = new Item[] { new Item("foo", 0, 0) };
         GildedRose app = new GildedRose(items);
         app.updateQuality();
-        assertEquals("fixme", app.items[0].name);
+        assertTrue(app.items[0].quality >= 0);
+    }
+
+    @Test
+    void qualityIsNeverMoreThan50() {
+        Item[] items = new Item[] { new Item("Aged Brie", 0, 50) };
```
Comment (mrfishball) [inline src/test/java/com/gildedrose/GildedRoseTest.java]: Question: What happens if you do:
```Java
Item[] items = new Item[] { new Item("Aged Brie", 0, 100) };
```

File: src/test/java/com/gildedrose/GildedRoseTest.java [hunk 3]
```diff
@@ -3,15 +3,105 @@
...
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 class GildedRoseTest {
 
     @Test
-    void foo() {
+    void qualityIsNeverNegative() {
         Item[] items = new Item[] { new Item("foo", 0, 0) };
```
Comment (mrfishball) [inline src/test/java/com/gildedrose/GildedRoseTest.java]: Question: What happens if you do:
```Java
Item[] items = new Item[] { new Item("foo", 0, -10) };
```

File: src/test/java/com/gildedrose/GildedRoseTest.java [hunk 4]
```diff
@@ -3,15 +3,105 @@
...
+    @Test
+    void qualityIsNeverMoreThan50() {
+        Item[] items = new Item[] { new Item("Aged Brie", 0, 50) };
+        GildedRose app = new GildedRose(items);
+        app.updateQuality();
+        assertTrue(app.items[0].quality <= 50);
+    }
+
+    @Test
+    void qualityDegradesTwiceAsFastWhenPassedSellBy() {
```
Comment (mrfishball) [inline src/test/java/com/gildedrose/GildedRoseTest.java]: Questions: `degrades twice as fast...` does that apply to all item types?

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 1 above)
Comment (riccjohn) [inline src/test/java/com/gildedrose/GildedRoseTest.java]:
```suggestion
        Item[] items = new Item[] { new Item("FLAvorTown", 0, 50) };
```

🤘🍕 🍔 🌭 🤘

Comment (mrfishball) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: I assume it's `0`?

Suggestion: Might want to consider named params like:
```Java
new Item(name: "FLAvorTown", sellBy: 0, quantities: 50)
```

It's also not clear what the unit for `sellBy`. I assume it's the number of days? so `5` would be 5 days? In that case maybe something like`sellByDays` would be more appropriate. Correct me if I'm wrong.

File: src/test/java/com/gildedrose/GildedRoseTest.java [hunk 5]
```diff
@@ -3,15 +3,105 @@
...
+    void legendaryItemDetails_ShouldNeverUpdate() {
+        Item[] items = new Item[] { new Item("Sulfuras, Hand of Ragnaros", 7, 80) };
+        GildedRose app = new GildedRose(items);
+        app.updateQuality();
+        assertEquals(80, app.items[0].quality);
+        assertEquals(7, app.items[0].sellIn);
+    }
+
+    @Test
+    void backstagePassesQualityIncreasesBy1_MoreThan10DaysAway() {
```
Comment (riccjohn) [inline src/test/java/com/gildedrose/GildedRoseTest.java]: Question: I don't know this testing framework, but is it possible to nest tests together? For example, could you have a section just for all the backstage pass tests so all that logic is together?

File: src/test/java/com/gildedrose/GildedRoseTest.java [hunk 6]
```diff
@@ -3,15 +3,105 @@
...
+    void updateNormalItem_QualityShouldReduceBy1() {
+        Item[] items = new Item[] { new Item("FLAvorTown", 7, 50) };
+        GildedRose app = new GildedRose(items);
+        app.updateQuality();
+        assertEquals(49, app.items[0].quality);
+    }
+
+    @Test
+    void updateNonLegendaryItem_SellByWillReduceBy1() {
+        Item[] items = new Item[] { new Item("FLAvorTown", 7, 50) };
```
Comment (mrfishball) [inline src/test/java/com/gildedrose/GildedRoseTest.java]: Questions: Should `FLAvorTown` be `FlavorTown` ?

**Comment by jonmanahan** [comment]
> How do I run this locally. What prerequisite software do I need installed on my machine, if any?

Addressed in [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/12/commits/497fbef597f358388bbc0fd1c9f99052347682f5) @elijahsullivan

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 1 above)
Comment (jonmanahan) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: Addressed in [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/12/commits/d638c0d71a1bf497cbb21dc0c3b8a78e9d2bbfde)

We noticed that the variable is set to `sellIn` within the `Item` class. So we did need to update the test names to reflect the proper variable of `sellIn` but in regards to your comment we do agree that specifying the unit would be more accurate. 

So we extracted the arguments into local variables that are now passed as arguments. And added "days" to the `sellIn` parameter. 

If you are not comfortable with the parameter name being `sellIn` we would need to go in and make a change to the `Item` class which is off limits at this point.

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 2 above)
Comment (jonmanahan) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: It looks like the previous developers did not add any safeguards around this use case, if any item is added to the inventory with a quality value over 50, the system will allow the items addition but its quality will never change.

So in this case, Aged Brie will remain with a quality of 100 no matter how many days pass.

We have added a bug ticket #13 to keep track of this issue and we will go over the requirements of this during the next IPM.

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 1 above)
Comment (mrfishball) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: I prefer refactoring it to use named params instead because it's much more explicit and has fewer indirections.

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 3 above)
Comment (jonmanahan) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: Same as above
> It looks like the previous developers did not add any safeguards around this use case

We have added a bug ticket #14 to keep track of this issue and we will go over the requirements of this during the next IPM.

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 4 above)
Comment (jonmanahan) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: That is a good point, it does not.

We have changed this test name to specify that this test is for normal items 

Addressed in [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/12/commits/15226d7abb12c26aa110be58e28a44f4f95f6478)

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 6 above)
Comment (jonmanahan) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: We were putting some extra UMPH in FLAvorTown, but we have come to our senses and reluctantly renamed it to FlavorTown

Comment (mrfishball) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: haha I see. I just wanted to make sure it's intentional and not some typo.

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 4 above)
Comment (mrfishball) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: I'd also suggest grouping tests by item type like:
```Java
@Test
void normalItems_qualityDegradesTwiceAsFastWhenPassSellBy() ...

void normalItems_update_qualityShouldReduceBy1()..

void legendaryItem_....

void backstagePasses_....

void agedBrie_....
```

Comment (jonmanahan) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: Addressed in [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/12/commits/3533d506a9d48299abad8a5061059d36e941e702)

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 5 above)
Comment (riccjohn) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: I think this is addressed in commit https://github.com/jramnani/GildedRoseJavaLLC/pull/12/commits/3533d506a9d48299abad8a5061059d36e941e702

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 4 above)
Comment (mrfishball) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: Very nice!

**Review by mrfishball — APPROVED** [review]
Left an unblocking comment about refactoring the `Item` class to use named params otherwise looks good to me.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Gilded Rose Inventory System
+
+## Description
+Welcome to the Gilded Rose Inventory System, this system currently keeps track of our various items sell by date and quality.
+
+## Prerequisites
+- Java
+  - To check your local Java version enter `java --version` in the command line.
```
Comment (riccjohn) [inline README.md:8]: Question: What version of java do we need to run this?

File: README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    - **If not installed** Follow the instruction details in this link -> [Java](https://java.com/en/download/)
+- Gradle
+  - To check if you have gradle running locally enter `gradle -v` in the command line
```
Comment (riccjohn) [inline README.md:11]: Question: What version of grade is required to run this?

**Comment by jonmanahan** [comment]
> Left an unblocking comment about refactoring the `Item` class to use named params otherwise looks good to me.

Thank you for review Steven 😄, the only thing about modifying the `Item` class is that I believe this is a file the client has requested/required us not to modify, just want to clarify this! Again appreciate the review and comments!

File: README.md (hunk 7 above)
Comment (jonmanahan) [reply README.md:8]: Addressed in [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/12/commits/5bf0a9e8200d17ab93f51f2c05267b29b89f26f3)

File: README.md (hunk 8 above)
Comment (jonmanahan) [reply README.md:11]: Addressed in [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/12/commits/5bf0a9e8200d17ab93f51f2c05267b29b89f26f3)

**Review by riccjohn — APPROVED** [review]

File: src/test/java/com/gildedrose/GildedRoseTest.java (hunk 1 above)
Comment (ikarabulut) [reply src/test/java/com/gildedrose/GildedRoseTest.java]: Marking resolved as this is not a blocker. We can talk about this approach during our next IPM if that is ok with you.

**Comment by ikarabulut** [comment]
Merging #11 PR to `trunk`
